pkg/platform/kube/platform.go

```diff
@@ -1260,3 +1272,47 @@ func (p *Platform) getDefaultServiceType() (v1.ServiceType, error) {
 		return "", errors.New("Not a valid configuration instance")
 	}
 }
+
+func (p *Platform) validateAPIGatewayFunctionsHaveNoIngresses(platformAPIGateway *platform.APIGatewayConfig) error {
+
+	// check ingresses on every upstream function
+	for _, upstream := range platformAPIGateway.Spec.Upstreams {
+		function, err := p.GetFunctions(&platform.GetFunctionsOptions{
+			Namespace: platformAPIGateway.Meta.Namespace,
+			Name: upstream.Nucliofunction.Name,
+		})
+		if err != nil {
+			return errors.New("Failed to get function")
+		}
+		if len(function) == 0 {
+
+			// if such function doesn't exist, just skip - because it doesn't have ingresses for sure
+			continue
+		}
+
+		ingresses := functionconfig.GetIngressesFromTriggers(function[0].GetConfig().Spec.Triggers)
+		if len(ingresses) > 0 {
+			return nuclio.NewErrConflict(fmt.Sprintf("Api gateway upstream function: %s must not have an ingress", upstream.Nucliofunction.Name))
+		}
+	}
+
+	return nil
+}
+
+// validate that a function is not exposed inside http triggers, while it is also exposed by an api gateway
+// this is done to prevent the nginx bug, where it is not working properly when the same service is exposed more than once
+// (specifically when it has canary ingresses)
+func (p *Platform) validateFunctionNoIngressAndAPIGateway(functionConfig *functionconfig.Config) error {
+	ingresses := functionconfig.GetIngressesFromTriggers(functionConfig.Spec.Triggers)
+	if len(ingresses) > 0 {
+		functionToAPIGateways, err := p.generateFunctionToAPIGatewaysMapping(functionConfig.Meta.Namespace)
```

Comment:
Im afraid that these validations are become heavier and heavier, especially where you "list" CRDs
assuming you have small amount - that is somehow acceptable,
but when moving to few hundreds, this should be optimized so consider optimizing this function to list.

There should be a function that get specific function apigateways using by labels, that way, you will reduce the `list` noise by N (amount of functions)

Reply:
Good idea, I'll just add labels for every api-gateway with its functions, and because it is a CRD we can edit those labels every time we update this api-gateway

I'll do this on a different PR